### PR TITLE
Bf16 pr

### PR DESCRIPTION
**Pull request by asleepzzz**
fix bf16 fwd pr and pv,
pr:
./bin/mlir-miopen-driver -p=false  -fil_layout=kcyx -in_layout=nchw -out_layout=nkhw -batchsize=16 -in_channels=16 -out_channels=64 -in_h=31 -in_w=31 -fil_h=1 -fil_w=1 --dilation_h=1 --dilation_w=1 --padding_h=0 --padding_w=0 --conv_stride_h=1 --conv_stride_w=1 -pr -ph  -c -x2 -t bf16

pv
./bin/mlir-miopen-driver -p=false  -fil_layout=kcyx -in_layout=nchw -out_layout=nkhw -batchsize=16 -in_channels=16 -out_channels=64 -in_h=31 -in_w=31 -fil_h=1 -fil_w=1 --dilation_h=1 --dilation_w=1 --padding_h=0 --padding_w=0 --conv_stride_h=1 --conv_stride_w=1 -pv  -c -x2 -t bf16 | ./bin/mlir-rocm-runner --shared-libs=./lib/librocm-runtime-wrappers.so,./lib/libmlir_runner_utils.so --entry-point-result=void

### CODE REVIEW

**Comment by whchung** [comment]
@asleepzzz Please make sure `check-mlir` pass before you submit the PR.

```
  MLIR :: mlir-miopen-driver/populate_host_convert.mlir
  MLIR :: mlir-miopen-driver/populate_host_print.mlir
```

**Review by whchung — CHANGES_REQUESTED** [review]
Relevant unit tests are not changed.

File: mlir/test/mlir-miopen-driver/populate_host_convert.mlir [hunk 1]
```diff
@@ -1,6 +1,6 @@
 // RUN: mlir-miopen-driver -p -ph -pr | FileCheck %s --check-prefix=F32
 // RUN: mlir-miopen-driver -p -ph -pr -t f16 | FileCheck %s --check-prefix=F16
-// RUN: mlir-miopen-driver -p -ph -pr -t bf16 | FileCheck %s --check-prefix=BF16
+// FIXME: BF16 use another function, so skip it
```
Comment (whchung) [inline mlir/test/mlir-miopen-driver/populate_host_convert.mlir]: remove FIXME.

File: mlir/test/mlir-miopen-driver/populate_host_convert.mlir [hunk 2]
```diff
@@ -40,23 +40,3 @@
 // F16-NEXT:   }
 // F16-NEXT: }
 // F16-NEXT: return
-
-// BF16: func @convert_result([[SOURCE:%[a-zA-Z_0-9]+]]: memref<[[N:[0-9]+]]x[[K:[0-9]+]]x[[HO:[0-9]+]]x[[WO:[0-9]+]]x[[TYPE:[a-zA-Z0-9]+]]>, [[DEST:%[a-zA-Z_0-9]+]]: memref<[[N]]x[[K]]x[[HO]]x[[WO]]x[[PRINT_TYPE:[a-zA-Z0-9]+]]>)
```
Comment (whchung) [inline mlir/test/mlir-miopen-driver/populate_host_convert.mlir:44]: add test to check bf16-specific conversion routine is present.

File: mlir/test/mlir-miopen-driver/populate_host_convert.mlir (hunk 1 above)
Comment (whchung) [reply mlir/test/mlir-miopen-driver/populate_host_convert.mlir]: and re-enable BF16 RUN command.

File: mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp [hunk 3]
```diff
@@ -1086,14 +1086,36 @@ static LogicalResult populateHostHarnessLogic(
...
   if (printResultTensor.getValue()) {
-    // Emit type conversion routine to convert every element to f32.
-    mlir::Value printHostValue = printHostAllocOp;
-    auto convertResultFuncOp = createConvertResult(
-        module, builder, resultOriginalCpuType, printMemRefType);
-    auto convertResultCallOp = builder.create<CallOp>(
-        builder.getUnknownLoc(), convertResultFuncOp,
-        ValueRange{resultOriginalCpuValue, printHostAllocOp});
-    block->push_back(convertResultCallOp);
+    if (dataType == builder.getIntegerType(16)) {
```
Comment (whchung) [inline mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp]: comment this path is for bf16(i16).

File: mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp [hunk 4]
```diff
@@ -1301,10 +1312,9 @@ static LogicalResult populateValidationLogic(
   };
 
   mlir::AllocOp gpuResults;
-  mlir::AllocOp gpuResultsBf16;
   if (operation.getValue() == "conv2d") {
     if (builder.getIntegerType(16) == dataType) {
-      gpuResultsBf16 = getFloatDataFromBF16(outputMemRefCastOp);
+      gpuResults = getFloatDataFromBF16(outputMemRefCastOp);
```
Comment (whchung) [inline mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp]: logic for backward data and backward filter are missing.

File: mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp [hunk 5]
```diff
@@ -1269,27 +1291,16 @@ static LogicalResult populateValidationLogic(
...
 
     auto verifyUnkownSizeMemRefCastOp = builder.create<MemRefCastOp>(
         builder.getUnknownLoc(), verifyHostAllocOp, unknownSizeMemRefFloatType);
     block->push_back(verifyUnkownSizeMemRefCastOp);
 
-    auto cpuMemConvertOp = FuncOp::create(
-        builder.getUnknownLoc(), "mcpuMemBF16ConvertFloat",
-        builder.getFunctionType(
-            {fourDimUnknownSizeMemRefType, unknownSizeMemRefFloatType}, {}));
+    auto cpuMemConvertOp = makeFuncDecl(
```
Comment (whchung) [inline mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp:1301]: is this `cpuMemConvertOp` really needed for all data types? line 1099 seems to have it already for bf16(i16) path.

File: mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp [hunk 6]
```diff
@@ -1526,12 +1532,15 @@ static LogicalResult populateValidationLogic(
...
   block->push_back(filterHostDeallocOp);
   block->push_back(inputHostDeallocOp);
   block->push_back(outputHostDeallocOp);
-  block->push_back(cpuResultsDeallocOp);
+
+  if (dataType != builder.getIntegerType(16)) { // i16 dealloc in below
+    auto cpuResultsDeallocOp =
+        builder.create<DeallocOp>(builder.getUnknownLoc(), cpuResults);
+    block->push_back(cpuResultsDeallocOp);
+  }
```
Comment (whchung) [inline mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp]: use `if / else` instead of 2 `if` blocks.

**Review by whchung — CHANGES_REQUESTED** [review]
A couple more minor changes and the PR is good to be taken.

File: mlir/test/mlir-miopen-driver/populate_host_convert.mlir (hunk 1 above)
Comment (asleepzzz) [reply mlir/test/mlir-miopen-driver/populate_host_convert.mlir]: but I don't use convert function when bf16

File: mlir/test/mlir-miopen-driver/populate_host_convert.mlir (hunk 2 above)
Comment (asleepzzz) [reply mlir/test/mlir-miopen-driver/populate_host_convert.mlir:44]: we use extern function to convert , the test is in populate_host_print.mlir

File: mlir/test/mlir-miopen-driver/populate_host_convert.mlir (hunk 1 above)
Comment (asleepzzz) [reply mlir/test/mlir-miopen-driver/populate_host_convert.mlir]: test added

File: mlir/test/mlir-miopen-driver/populate_host_convert.mlir (hunk 2 above)
Comment (asleepzzz) [reply mlir/test/mlir-miopen-driver/populate_host_convert.mlir:44]: test added

File: mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp (hunk 4 above)
Comment (asleepzzz) [reply mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp]: fixed

File: mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp (hunk 3 above)
Comment (asleepzzz) [reply mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp]: fixed

File: mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp (hunk 6 above)
Comment (asleepzzz) [reply mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp]: fixed

**Review by whchung — APPROVED** [review]
LGTM

**Pull request merged**